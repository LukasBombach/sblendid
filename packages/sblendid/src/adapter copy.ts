import { inherits } from "util";
import promisedEvent from "p-event";
import Bindings, {
  Events,
  EventName as Event,
  EventListener as Listener,
  EventParameters as Params
} from "../types/bindings";
import Peripheral from "./peripheral";

export type Action = () => void | Promise<void>;
export type When<E extends Event> = () => Promise<Params<E>>;
export type End = () => void | Promise<void>;
export type Condition<E extends Event> = Listener<E> | Params<E> | Params<E>[0];

export type DiscoverListener = (peripheral: Peripheral) => Promise<void> | void;

type ListenerMap = Map<Function, Listener<Event>>;

// const mapEventToPeripheral = (...p: DiscoverParams) => new Peripheral(this.adapter, ...p);

interface MappedListeners {
  discover: (peripheral: Peripheral) => Promise<void> | void;
}

interface EventMappers {
  discover: (...p: Params<"discover">) => Peripheral;
}

type AnyListener<E extends Event> = Listener<E> | MappedListener<E>;

type MappedFunction = Function; // Type "Function"
export type EventOptions = EventOptionsObject | MappedFunction;
export interface EventOptionsObject {
  map?: MappedFunction | boolean;
}

export default class Adapter extends Bindings {
  private listenerMap: ListenerMap = new Map();

  private eventMappers: EventMappers = {
    discover: (...p: Params<"discover">) => new Peripheral(this, ...p)
  };

  private constructor() {
    super();
  }

  public static withBindings(bindings: Bindings): Bindings & Adapter {
    class BoundAdapter {}
    inherits(BoundAdapter, bindings.constructor);
    inherits(BoundAdapter, Adapter);
    return new BoundAdapter() as Bindings & Adapter;
  }

  // todo bad typecast
  on<E extends Event>(event: E, listener: Listener<E>): void;
  on<E extends Event>(event: E, listener: Listener<E>, options: EventOptions): void;
  on<E extends Event>(event: E, listener: Listener<E>, options?: EventOptions): void {
    const mapFunction = this.getMapFunction(event, options);
    if (!mapFunction) return super.on(event, listener);
    if (this.listenerMap.get(mapFunction)) return;
    const lowLevelListener = ((...args: Params<E>) => mapFunction(...args)) as Events[E];
    this.listenerMap.set(mapFunction, lowLevelListener);
    super.on<E>(event, lowLevelListener);
  }

  // todo bad typecast
  once<E extends Event>(event: E, listener: Listener<E>, options?: EventOptions): void {
    const mapFunction = this.getMapFunction(event, options);
    if (!mapFunction) return super.on(event, listener);
    if (this.listenerMap.get(mapFunction)) return;
    const lowLevelListener = ((...args: Params<E>) => mapFunction(...args)) as Events[E];
    this.listenerMap.set(mapFunction, lowLevelListener);
    super.once<E>(event, lowLevelListener);
  }

  // todo bad typecast
  off<E extends Event>(event: E, listener: Listener<E>, options?: EventOptions): void {
    const mapFunction = this.getMapFunction(event, options);
    if (!mapFunction) return super.on(event, listener);
    const lowLevelListener = this.listenerMap.get(mapFunction) as Events[E];
    if (lowLevelListener) this.listenerMap.delete(mapFunction);
    super.off<E>(event, lowLevelListener);
  }

  async run<E extends Event>(action: Action, when: When<E>, end?: End): Promise<Params<E>> {
    const [eventParameters] = await Promise.all([when(), action()]);
    if (end) await end();
    return eventParameters;
  }

  // todo fucking any, same reason as for the typecast above
  public when<E extends Event>(event: E, filter?: Condition<E>): Promise<Params<E>> {
    return promisedEvent<E, Params<E>>(this as any, event, { filter, multiArgs: true } as any);
  }

  // todo this should return the low level listener as well
  private getMapFunction(event: Event, options?: EventOptions): MappedFunction | undefined {
    if (typeof options === "undefined") return undefined;
    if (typeof options === "function") return options;
    if (typeof options.map === "function") return options.map;
    if (options.map === true) return this.eventMaps[event];
    return;
  }
}
