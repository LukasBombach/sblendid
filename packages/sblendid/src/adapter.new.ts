import { EventEmitter } from "events";
import {
  Bindings,
  EventName as Event,
  EventListener as Listener,
  EventParameters as Params
} from "sblendid-bindings-macos";
import Peripheral from "./peripheral";

export interface Converters {
  discover: (...args: Params<"discover">) => Peripheral;
}

export type Converters2 = Record<Event, () => any>;

export interface EventOptions {
  convert: boolean;
}

export type ConvListener<E extends keyof Converters2> = Converters2[E];
export type AnyListener<E extends Event> = Listener<E> | ConvListener<E>;

export default class Adapter {
  private bindings: Bindings;
  private convertEmitter: EventEmitter;

  private converters: Converters = {
    discover: (...args: Params<"discover">) =>
      this.convertEmitter.emit("discover", Peripheral.fromDiscover(this, args))
  };

  constructor(bindings: Bindings = new Bindings()) {
    this.convertEmitter = new EventEmitter();
    this.bindings = bindings;
    this.convertEmitter.on("newListener", this.onNewConvertListener.bind(this));
    this.convertEmitter.on("removeListener", this.onRemoveConvertListener.bind(this));
  }

  public on<E extends Event>(event: E, listener: Listener<E>): void;
  public on<E extends Event>(event: E, listener: ConvListener<E>, options: EventOptions): void;
  public on<E extends Event>(event: E, listener: AnyListener<E>, options?: EventOptions): void {
    if (options && options.convert) {
      this.convertEmitter.on(event, listener);
    } else {
      this.bindings.on(event, listener);
    }
  }

  public off<E extends Event>(event: E, listener: AnyListener<E>): void {
    if (this.convertEmitter.listeners(event).includes(listener)) {
      this.convertEmitter.off(event, listener);
    } else {
      this.bindings.off(event, listener);
    }
  }

  private onNewConvertListener(event: Event): void {
    if (this.convertEmitter.listenerCount(event) > 0) return;
    this.bindings.on(event, this.converters[event]);
  }

  private onRemoveConvertListener(event: Event): void {
    if (this.convertEmitter.listenerCount(event) > 0) return;
    this.bindings.off(event, this.converters[event]);
  }
}
