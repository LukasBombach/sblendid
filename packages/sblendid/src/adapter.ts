import { inherits } from "util";

import { Bindings, EventName, EventListener, EventParameters } from "sblendid-bindings-macos";
import promisedEvent from "p-event";

export type Action = () => Promise<void> | void;
export type When<E extends EventName> = () => Promise<EventParameters<E>>;
export type End = () => Promise<void> | void;

export type Condition<E extends EventName> =
  | EventListener<E>
  | EventParameters<E>
  | EventParameters<E>[0];

export default class Adapter extends Bindings {
  /* public static withBindings(bindings: Bindings): Bindings & Adapter {
    class BoundAdapter {}
    inherits(BoundAdapter, bindings.constructor);
    inherits(BoundAdapter, Adapter);
    return new BoundAdapter() as Bindings & Adapter;
  } */

  async run<E extends EventName>(
    action: Action,
    when: When<E>,
    end?: End
  ): Promise<EventParameters<E>> {
    const [eventParameters] = await Promise.all([when(), action()]);
    if (end) await end();
    return eventParameters;
  }

  // todo fucking any
  public when<E extends EventName>(event: E, filter?: Condition<E>): Promise<EventParameters<E>> {
    const multiArgs = true;
    //const filter = condidion
    return promisedEvent<E, EventParameters<E>>(this as any, event, { filter, multiArgs } as any);
  }
}
