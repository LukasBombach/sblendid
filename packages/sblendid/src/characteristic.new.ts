export interface Converter<T = Buffer> {
  uuid: CUUID;
  decode?: (value: Buffer) => Promish<T>;
  encode?: (value: T) => Promish<Buffer>;
}

export interface Properties {
  read?: boolean;
  write?: boolean;
  notify?: boolean;
}

export type Listener<C extends Converter<any>> = (
  value: T
) => Promise<void> | void;

export default class Characteristic<C extends Converter<any>> {
  public uuid: CUUID;
  public properties: Properties;
  private converter?: C;

  constructor(uuid: CUUID, properties: Properties = {}, converter?: C) {
    this.uuid = uuid;
    this.properties = properties;
    this.converter = converter;
  }

  public async read(): Promise<T> {}

  public async write(value: T, withoutResponse?: boolean): Promise<void> {}

  public async on(event: "notify", listener: Listener<T>): Promise<void> {}

  public async off(event: "notify", listener: Listener<T>): Promise<void> {}
}
