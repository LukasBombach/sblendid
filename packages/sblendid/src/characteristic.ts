import { NobleCharacteristic, Params } from "./bindings";
import Adapter from "./adapter";
import Service from "./service";
import { EventEmitter } from "events";

const defaultProperties: Properties = {
  read: false,
  write: false,
  notify: false
};

export default class Characteristic<T = Buffer> {
  public uuid: CUUID;
  public properties?: Properties;
  public service: Service;
  private adapter: Adapter;
  private converter?: Converter<T>;
  private eventEmitter = new EventEmitter();
  private isNotifying = false;

  public static fromNoble<T>(
    service: Service,
    noble: NobleCharacteristic,
    converter?: Converter<T>
  ): Characteristic<T> {
    const properties = Object.assign({}, defaultProperties);
    for (const name of noble.properties) properties[name] = true;
    return new Characteristic(service, noble.uuid, converter, properties);
  }

  constructor(
    service: Service,
    uuid: CUUID,
    converter?: Converter<T>,
    properties?: Properties
  ) {
    this.service = service;
    this.uuid = uuid;
    this.converter = converter;
    this.properties = properties;
    this.adapter = service.adapter;
  }

  public async read(): Promise<T> {
    const buffer = await this.dispatchRead();
    return await this.decode(buffer);
  }

  public async write(value: T): Promise<void> {
    const buffer = await this.encode(value);
    await this.dispatchWrite(buffer);
  }

  public async on(event: "notify", listener: Listener<T>): Promise<void> {
    this.eventEmitter.on(event, listener);
    if (!this.isNotifying) await this.startNotifing();
  }

  public async off(event: "notify", listener: Listener<T>): Promise<void> {
    if (!this.eventEmitter.listenerCount("notify")) await this.stopNotifing();
    this.eventEmitter.off(event, listener);
  }

  private async decode(buffer: Buffer): Promise<T> {
    if (!this.converter || !this.converter.decode) return buffer as any;
    return await this.converter.decode(buffer);
  }

  private async encode(value: T): Promise<Buffer> {
    if (!this.converter || !this.converter.encode) return value as any;
    return await this.converter.encode(value);
  }

  private async startNotifing(): Promise<void> {
    if (this.isNotifying) return;
    this.adapter.on("read", this.onNotify.bind(this));
    this.isNotifying = await this.notify(true);
  }

  private async stopNotifing(): Promise<void> {
    if (!this.isNotifying) return;
    this.adapter.off("read", this.onNotify.bind(this));
    this.isNotifying = await this.notify(false);
  }

  private async onNotify(...params: Params<"read">): Promise<void> {
    const [pUuid, sUuid, cUuid, data, isNfy] = params;
    if (!this.isThis(pUuid, sUuid, cUuid)) return;
    this.eventEmitter.emit("notify", await this.decode(data), { isNfy });
  }

  private isThis(pUuid: string, sUuid: SUUID, cUuid: CUUID): boolean {
    return this.getUuids().every((v, i) => v === [pUuid, sUuid, cUuid][i]);
  }

  private getUuids(): [string, SUUID, CUUID] {
    return [this.service.peripheral.uuid, this.service.uuid, this.uuid];
  }

  private async dispatchRead(): Promise<Buffer> {
    const [pUuid, sUuid, uuid] = this.getUuids();
    return await this.adapter.run<"read", Buffer>(
      () => this.adapter.read(pUuid, sUuid, uuid),
      () => this.adapter.when("read", (p, s, c) => this.isThis(p, s, c)),
      ([, , , buffer]) => buffer
    );
  }

  private async dispatchWrite(value: Buffer): Promise<void> {
    const [pUuid, sUuid, uuid] = this.getUuids();
    await this.adapter.run<"write">(
      () => this.adapter.write(pUuid, sUuid, uuid, value, true),
      () => this.adapter.when("write", (p, s, c) => this.isThis(p, s, c))
    );
  }

  private async notify(notify: boolean): Promise<boolean> {
    const [pUuid, sUuid, uuid] = this.getUuids();
    return await this.adapter.run<"notify", boolean>(
      () => this.adapter.notify(pUuid, sUuid, uuid, notify),
      () => this.adapter.when("notify", (p, s, c) => this.isThis(p, s, c)),
      ([, , , state]) => state
    );
  }
}
