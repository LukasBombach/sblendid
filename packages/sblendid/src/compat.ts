import BuiltinModule from "module";

type AliasModule = Function & { _resolveFilename: ResolveFilename };
type ResolveFilename = (
  request: string,
  parentModule: BuiltinModule,
  isMain: boolean
) => any;

export default class AliasNoble {
  private static replacements: [string[], string][] = [
    [["noble", "noble/lib/noble"], "sblendid/compat"]
  ];
  private static moduleRef: AliasModule = AliasNoble.getModule();
  private static originalResolve: ResolveFilename = AliasNoble.getOriginalResolve();

  public static shim(): void {
    const shim: ResolveFilename = (...args) => AliasNoble.resolveShim(...args);
    AliasNoble.moduleRef._resolveFilename = shim;
  }

  public static unshim(): void {
    AliasNoble.moduleRef._resolveFilename = AliasNoble.originalResolve;
  }

  private static resolveShim(
    request: string,
    parentModule: BuiltinModule,
    isMain: boolean
  ): any {
    for (const [matches, replacement] of AliasNoble.replacements) {
      if (matches.includes(request)) request = replacement;
    }
    return AliasNoble.originalResolve(request, parentModule, isMain);
  }

  private static getModule(): AliasModule {
    if (module.constructor.length > 1) return module.constructor as any;
    return BuiltinModule as any;
  }

  private static getOriginalResolve(): ResolveFilename {
    return AliasNoble.getModule()._resolveFilename;
  }
}
