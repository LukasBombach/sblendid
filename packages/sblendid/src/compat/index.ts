export { default } from "./noble";
export { default as aliasNoble } from "./aliasNoble";
