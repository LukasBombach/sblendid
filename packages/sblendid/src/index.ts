/// <reference path="./types.d.ts" />

export { default } from "./sblendid";
export { default as Peripheral } from "./peripheral";
export { default as Service } from "./service";
export { default as Characteristic } from "./characteristic";

export { Value, Listener } from "./service";
export { Converter } from "./characteristic";
