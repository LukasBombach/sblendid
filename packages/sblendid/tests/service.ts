import Service from "../src/service";
import Sblendid from "../src/sblendid";
import Peripheral from "../src/peripheral";
import { Converter } from "../src/characteristic";

describe("Service", () => {
  const name = "Find Me";
  const deviceInfoUUID = "180a";
  const manufacturerUUID = "2a29";

  let peripheral: Peripheral;
  let services: Service<any>[];

  let writableSUUID: SUUID;
  let writableCUUID: CUUID;
  let notifyableSUUID: SUUID;
  let notifyableCUUID: CUUID;
  let writeConverters: { message: Converter<string> };
  let notifyConverters: { message: Converter<string> };

  const converters = {
    manufacturer: {
      uuid: "2a29",
      decode: (buffer: Buffer) => buffer.toString()
    }
  };

  beforeAll(async () => {
    peripheral = await Sblendid.connect(name);
    // todo this needs to be done before services can be used directly
    // todo this must be done automatically by the service class
    services = await peripheral.getServices();
    // todo this is hands down the worst code I have ever seen
    while (writableSUUID === undefined && notifyableSUUID === undefined) {
      for (const service of services) {
        const characteristics = await service.getCharacteristics();
        for (const { uuid, properties } of characteristics) {
          if (properties.write) {
            writableSUUID = service.uuid;
            writableCUUID = uuid;
          }
          if (properties.notify) {
            notifyableSUUID = service.uuid;
            notifyableCUUID = uuid;
          }
        }
      }
    }

    writeConverters = {
      message: {
        uuid: writableCUUID,
        decode: (buffer: Buffer) => buffer.toString(),
        encode: (message: string) => Buffer.from(message, "utf8")
      }
    };
    notifyConverters = {
      message: {
        uuid: notifyableCUUID,
        decode: (buffer: Buffer) => buffer.toString(),
        encode: (message: string) => Buffer.from(message, "utf8")
      }
    };
  }, 10000);

  afterAll(async () => {
    await peripheral.disconnect();
  });

  it("can be instantiated without converters", async () => {
    expect(() => new Service(peripheral, deviceInfoUUID)).not.toThrow();
  }, 10000);

  it("can be instantiated with converters", async () => {
    expect(
      () => new Service(peripheral, deviceInfoUUID, converters)
    ).not.toThrow();
  }, 10000);

  it("reads a characteristic using its UUID", async () => {
    const deviceInfoService = new Service(peripheral, deviceInfoUUID);
    const buffer = await deviceInfoService.read(manufacturerUUID);
    expect(buffer).toBeInstanceOf(Buffer);
    expect(buffer.toString()).toMatch(/.+/);
  }, 10000);

  it("reads a characteristic using a converter name", async () => {
    const deviceInfoService = new Service(
      peripheral,
      deviceInfoUUID,
      converters
    );
    const value = await deviceInfoService.read("manufacturer");
    expect(typeof value).toBe("string");
    expect(value).toMatch(/.+/);
  }, 10000);

  it("writes a characteristic using its UUID", async () => {
    const service = new Service(peripheral, writableSUUID);
    await expect(
      service.write(writableCUUID, Buffer.from("message"))
    ).resolves.toBe(undefined);
  }, 10000);

  it("writes a characteristic using a converter name", async () => {
    const service = new Service(peripheral, writableSUUID, writeConverters);
    await expect(service.write("message", "message")).resolves.toBe(undefined);
  }, 10000);

  it("writes a characteristic using its UUID without response", async () => {
    const service = new Service(peripheral, writableSUUID);
    await expect(
      service.write(writableCUUID, Buffer.from("message"), true)
    ).resolves.toBe(undefined);
  }, 10000);

  it("writes a characteristic using a converter name without response", async () => {
    const service = new Service(peripheral, writableSUUID, writeConverters);
    await expect(service.write("message", "message", true)).resolves.toBe(
      undefined
    );
  }, 10000);

  it("can listen to notifications using a UUID", async () => {
    const service = new Service(peripheral, notifyableSUUID);
    await expect(service.on(notifyableCUUID, () => {})).resolves.toBe(
      undefined
    );
  }, 10000);

  it("can listen to notifications using a converter name", async () => {
    const service = new Service(peripheral, notifyableSUUID, notifyConverters);
    await expect(service.on("message", () => {})).resolves.toBe(undefined);
  }, 10000);

  it("can stop listening to notifications using a UUID", async () => {
    const service = new Service(peripheral, notifyableSUUID);
    const listener = () => {};
    service.on(notifyableCUUID, listener);
    await expect(service.off(notifyableCUUID, listener)).resolves.toBe(
      undefined
    );
  }, 10000);

  it("can stop listening to notifications using a converter name", async () => {
    const service = new Service(peripheral, notifyableSUUID, notifyConverters);
    const listener = () => {};
    service.on("message", listener);
    await expect(service.off("message", listener)).resolves.toBe(undefined);
  }, 10000);

  it("gets all avalable characteristics", async () => {
    const service = new Service(peripheral, deviceInfoUUID);
    const characteristics = await service.getCharacteristics();
    expect(characteristics.map(c => c.uuid).sort()).toMatchSnapshot();
  }, 10000);

  it("gets all avalable characteristics using converers", async () => {
    const service = new Service(peripheral, deviceInfoUUID, converters);
    const characteristics = await service.getCharacteristics();
    expect(characteristics.map(c => c.uuid).sort()).toMatchSnapshot();
  }, 10000);
});
