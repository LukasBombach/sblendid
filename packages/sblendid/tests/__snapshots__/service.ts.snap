// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service gets all avalable characteristics 1`] = `
Array [
  "2a24",
  "2a29",
]
`;

exports[`Service gets all avalable characteristics using converers 1`] = `
Array [
  "2a24",
  "2a29",
]
`;
